src/mds/MDCache.cc

```diff
@@ -13107,5 +13121,22 @@ void MDCache::handle_mdsmap(const MDSMap &mdsmap) {
     it = q.erase(it);
     in->maybe_export_pin();
   }
-}
 
+  /* Handle consistent hash ring changes when new mds is active. In the case where an mds is stopped (max_mds is decreased), the subtree migrations are handled as the Inodes are loaded */
+  if (mdsmap.get_max_mds() > oldmap.get_max_mds()) {
```

Comment:
Why not `!=`?

Reply:
Because the `<` case is handled differently. When max_mds is reduced the inodes (that were a part of the stopped mds) that get loaded in other MDSs get hashed again and sent elsewhere.

Reply:
Thanks for adding a comment. From this:

> In the case where an mds is stopped (max_mds is decreased), the subtree migrations are handled as the Inodes are loaded

but I still don't understand. The inodes are already loaded. If we reduce max_mds, then inodes on the stopping MDS will need migrated to their new home. Current that's rank 0 IIRC. It is fine to maintain that (lazy but simple) behavior but keep in mind that once the rank stops, rank0 should be migrating the ephemerally pinned subtrees as required. I don't see how/where that's done in this code.

Reply:
Ack. As it is now, the stopping MDS does not migrate the inodes that it handles (the inodes in its consistent hash list). The loaded inodes are dropped and once the MDS is stopped and there are only n-1 MDSs then when client requests come to the dir that was held by the now defunct nth MDS, rank 0 loads the metadata from RADOS, and
1) For ephemeral_random: Redoes the pinning probabilistically. May or may not get pinned again.
2) For ephemeral_distributed: Finds that the parent has ephemeral_distributed set and hence rehashes and redoes the pinning definitely.

I was worried about the delay in stopping time of the MDS, and other penalties if the migrations were done during the shutdown phase. Also in the current way, the loading is done when requests arrive and so is more dynamic(??). Do you think there is more penalties involved with reloading from RADOS, rather than migrations?

Reply:
> Ack. As it is now, the stopping MDS does not migrate the inodes that it handles (the inodes in its consistent hash list). The loaded inodes are dropped and once the MDS is stopped and there are only n-1 MDSs then when client requests come to the dir that was held by the now defunct nth MDS, rank 0 loads the metadata from RADOS, and

Not necessarily, IIRC. The stopping MDS needs to migrate any metadata that clients have capabilities for.

In any case, the code needs to consider shrinking max_mds. Consider `max_mds` shrinking from `3` to `2`:

````
|    0    |    1    |    2    |
|      0       |       1      |
````

Rank 1 clearly needs to migrate some of the metadata it has to rank 0.

>     1. For ephemeral_random: Redoes the pinning probabilistically. May or may not get pinned again.

Again, only true if the metadata is unpinned (no capabilities are issued, the metadata is unreplicated, etc.).

>     2. For ephemeral_distributed: Finds that the parent has ephemeral_distributed set and hence rehashes and redoes the pinning definitely.

Ditto.

> I was worried about the delay in stopping time of the MDS, and other penalties if the migrations were done during the shutdown phase.

Migrations are unavoidable. Note that in `MDCache::shutdown_pass`, the MDS tries to trip everything unpinned. Anything leftover must be exported or (eventually) expired if dirty.

> Also in the current way, the loading is done when requests arrive and so is more dynamic(??). Do you think there is more penalties involved with reloading from RADOS, rather than migrations?

Migrations would usually be cheaper in any case as the metadata will usually be warmer and loaded into the other MDS cache in batch.

Reply:
Ack! I have changed it and added this logic in MDCache::shutdown_pass.

Reply:
No, it needs to be here too. All MDS ranks need to check their current ephemerally pinned inodes, not just the one shutting down.